2020/02/21/consolidate-multiple-sftp-accounts-on-one-master-account.md

```diff
@@ -0,0 +1,154 @@
+tags: ssh, sftp, shell
+---
+
+
+Recently, a data intensive client implemented a workflow process to generate various reports and insights from a list of facilities. Because a significant portion of these files contain sensitive data, they would need to strictly adhere to HIPAA compliance. Optimally, facilities should be able to transfer files securely and exclusively to our server. Arguably, the best method of achieving this is to create individual SSH File Transfer Protocol (SFTP) accounts for each source.
+
+
+### SFTP Account
+
+Private SFTP accounts were established for each facility and the data was received at a designated path. At these individual points of contact, a third party application picks up the data and processes further into the pipeline. The following demonstrates how SFTP accounts are developed and configured:
+
+
+* Create a user group for SFTP accounts
+
+```bash
+$ addgroup sftpusers
+```
+
+* Configure the following settings in sshd_config (this enables an SFTP account and defaults the location as the home path)
+
+```bash
+$ vi /etc/ssh/sshd_config
+
+Subsystem       sftp    internal-sftp
+
+Match Group sftpusers
+    ChrootDirectory /home/%u
+    AllowTCPForwarding no
+    X11Forwarding no
+    ForceCommand internal-sftp
+```
+
+* Restart SSH Server to apply changes
+
+```bash
+$ /etc/init.d/ssh restart
+```
+
+* Create an SFTP user account for a facility and place in a folder on the home path to receive data
+
+```bash
+useradd $USERNAME
+usermod -g sftpusers -s /usr/sbin/nologin $USERNAME
+mkdir -p /home/$USERNAME
+mkdir -p /home/$USERNAME/INPUT_PATH/
+chown -R root:root /home/$USERNAME
+```
+
+As the number of facilities increases, additional communication points are added. If the permissible amount of points is exceeded on the app by even one, it will not continue its service until more blocks are paid for. The implementation of this process will directly benefit our client.
+
+### Multiple Accounts Mounts to One Account
+
+The goal here is to point the data from many facilities to one location. Yet, using a single account and path for multiple sites’ data will result in a breach in security and privacy. However, mounting the receiving path of a facility’s data onto a single master account to then “mount point” with a unique facility name takes care of this issue. The process continues on to consolidate files from individual paths on a master account in one place where the application picks up messages for further processing.
+
+
+Both the SFTP accounts as well as master account should be attached to one group. This will permit individual accounts to write on the master account-mounted path. In turn, the master account can read files from the same location. This location now has administrative rights of both the SFTP user as well as shared group permission. Group permission of the mounted folder is set to sftpgroup and user permission is set to facility account.
```

Comment:
Are you sure that users can't see each others' uploads? It looks to me like, since they are each a member of the master group, if they can upload, they can also read others' uploads.

Reply:
The sftpusers and master are member of `sftpgroup` and sftpuser doesn't have master group. The sftpusers home directory is configured as their root directory (`ChrootDirectory /home/%u`), so they can't navigate backwards to /home to access others users.
```
$ sftp sftparup@sl64.endpoint.com
sftparup@sl64.endpoint.com's password: 
Connected to sftparup@sl64.endpoint.com.
sftp> ls
SFTP_ELR  
sftp> pwd
Remote working directory: /
```

Reply:
@selvait90 So you are saying that the isolation of each user's data from each other comes not from any file permissions, but rather from the chroot. Is that right?

I just want to be clear about what it is.

Reply:
@jonjensen 
Right, chroot provides isolation among sftpaccounts and file group permission allows master account to access all files across sftpaccounts.